egs/aishell/s10/chain/chain_loss.py

```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env python3
+
+# Copyright 2019 Mobvoi AI Lab, Beijing, China (author: Fangjun Kuang)
+# Apache 2.0
+
+import torch
+from torch.autograd import Function
+from torch.utils.dlpack import to_dlpack
+
+import kaldi
+import kaldi_pybind.chain as chain
+
+g_nnet_output_deriv_tensor = None
+g_xent_output_deriv_tensor = None
+
+
+class KaldiChainObjfFunction(Function):
+
+    @staticmethod
+    def forward(ctx, opts, den_graph, supervision, nnet_output_tensor,
+                xent_out_unused_tensor):
+        '''
+        opts:          Struct containing options
+
+        den_graph:     The denominator graph, derived from denominator fst.
+
+        supervision:   The supervision object, containing the supervision
+                       paths and constraints on the alignment as an FST.
+
+        nnet_output_tensor:   The output of the neural net; dimension must
+                              equal ((supervision.num_sequences * supervision.frames_per_sequence)
+                              by den_graph.NumPdfs()).  The rows are ordered
+                              as: all sequences for frame 0; all sequences for
+                              frame 1; etc.
+
+        xent_out_unused_tensor: It MUST have the same structure as `nnet_output_tensor`;
+                                it is passed but NOT used directly in this
+                                function; it is still needed because we need
+                                to do back propagation for it.
+        '''
+        global g_nnet_output_deriv_tensor
+        global g_xent_output_deriv_tensor
+
+        if g_nnet_output_deriv_tensor is None or g_nnet_output_deriv_tensor.shape != nnet_output_tensor.shape:
+            # preallocate GPU memory in PyTorch and it will be shared with Kaldi
+            g_nnet_output_deriv_tensor = torch.zeros_like(nnet_output_tensor)
+            g_xent_output_deriv_tensor = torch.zeros_like(
+                nnet_output_tensor).contiguous()
+            # (fangjun): Note that g_xent_output_deriv needs to be contiguous,
+            # i.e., stride == cols; otherwise, it will be re-allocated in Kaldi
+        else:
+            # (fangjun): we have to zero them out manually!
+            g_nnet_output_deriv_tensor.zero_()
+            g_xent_output_deriv_tensor.zero_()
+
+        # it contains [objf, l2_term, weight] and will be returned to the caller
+        objf_l2_term_weight_tensor = torch.zeros(3).float()
+
+        nnet_output = kaldi.CuSubMatrixFromDLPack(to_dlpack(nnet_output_tensor))
```

Comment:
I'm thinking it would be nice to have a shorter convenience mechanism for
`kaldi.CuSubMatrixFromDLPack(to_dlpack(nnet_output_tensor))`.
e.g.
`kaldi.PytorchToCuSubMatrix(x)`
?
what do you think?

Reply:
I agree.

For example, in some file `util.py`, we can write
```python
def from_tensor(tensor):
   m = kaldi.CuSubMatrixFromDLPack(to_dlpack(nnet_output_tensor))
   return m
```

Then we can use

```python
from util import from_tensor
nnet_output = from_tensor(nnet_output_tensor)
```